aether/python/avro/generation.py

```diff
@@ -0,0 +1,252 @@
+# Copyright (C) 2019 by eHealth Africa : http://www.eHealthAfrica.org
+#
+# See the NOTICE file distributed with this work for additional information
+# regarding copyright ownership.
+#
+# Licensed under the Apache License, Version 2.0 (the 'License');
+# you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from copy import copy
+from datetime import datetime, timedelta
+import inspect
+import json
+import random
+from string import ascii_lowercase
+from typing import Any, List, Mapping
+import uuid
+
+from spavro.schema import parse as parse_schema
+
+from aether.python.avro.schema import Node
+from aether.python.avro.tools import (
+    AvroValidationException,
+    validate,
+)
+from aether.python.utils import replace_nested
+
+'''
+    Utility Functions
+'''
+
+
+def guid():
+    return str(uuid.uuid4())
+
+
+def single_choice(choices):
+    # return one item from a list
+    return random.choice(choices)
+
+
+def subset(choices, max=None):
+    # return a random subset from some choices
+    if not max:
+        max = len(choices)
+    size = random.randint(1, max)
+    return random.sample(choices, size)
+
+
+def str_time_prop(start, end, fraction):
+    # get an iso string of a point a fraction
+    # between a start and end datetime
+    stime = datetime.fromisoformat(start)
+    etime = datetime.fromisoformat(end)
+    new_time = stime + fraction * (etime - stime)
+    return new_time.isoformat()
+
+
+'''
+    Callable From the Generator
+    (only uses kwargs, and might know how to use Nodes)
+'''
+
+
+def random_date(start=None, end=None, min_prop=0, max_prop=1):
+    prop = min_prop + random.random() * (max_prop - min_prop)
+    return str_time_prop(start, end, prop)
+
+
+def random_str(constant=None, max=32):
+    if constant:
+        return constant
+    return ''.join([
+        c for i in range(max) for c in random.choice(ascii_lowercase)
+    ])
+
+
+def random_numeric(min=0, max=1, cast=int):
+    return cast(random.random() * (max - min) + min)
+
+
+def select1(choices=None, node=None):
+    if not choices:
+        choices = [i.get('value') for i in node.__lookup]
+    return single_choice(choices)
+
+
+def select_many(choices=None, node=None, max=None, path=None):
+    if not choices:
+        choices = [i.get('value') for i in node.__lookup]
+    return subset(choices, max)
+
+
+class SampleGenerator(object):
```

Comment:
Could this class replace the `random_avro` utility declared in https://github.com/eHealthAfrica/aether-python-library/blob/master/aether/python/avro/tools.py#L96?

Reply:
I'm getting a lot of schema parse errors from Spavro when the SampleGenerator is trying to load them to validate the payload. Are all of those bits valid Avro?

Reply:
yes :thinking:

Reply:
```python
class TestAvroRandom(TestCase):

    def test_random_avro__null(self):
        self.assertIsNone(random_avro({'type': 'null'}))

    def test_random_avro__boolean(self):
        self.assertTrue(isinstance(random_avro({'type': 'boolean'}), bool))

    def test_random_avro__int(self):
        self.assertTrue(isinstance(random_avro({'type': 'int'}), int))

    def test_random_avro__long(self):
        self.assertTrue(isinstance(random_avro({'type': 'long'}), int))

    def test_random_avro__float(self):
        self.assertTrue(isinstance(random_avro({'type': 'float'}), float))

    def test_random_avro__double(self):
        self.assertTrue(isinstance(random_avro({'type': 'double'}), float))

    def test_random_avro__string(self):
        self.assertTrue(isinstance(random_avro({'type': 'string'}), str))
```

Reply:
It looks like we only handle things with a base type of record at the moment.

Reply:
Do you want to fix this in this PR or in another one?

I would be happy not having duplicated functionality but I can understand that this can wait.

Reply:
I've reused the random_avro functionality in the SampleGenerator class for any type that doesn't have an explicit override.